### ISSUE TITLE
Vim doesn't immediately apply `wincolor` to terminal window, only after pressing CTRL-L

### ISSUE THREAD

**Issue by bfrg** [vim/vim#5818]
**Describe the bug**
The `wincolor` of a terminal window is not applied immediately in Gvim. The color is set correctly only after pressing `CTRL-L`. The same can be observed in terminal Vim when `termguicolors` is enabled. It works as expected with `notermguicolors`.

_Note:_ When the terminal buffer is opened in a popup window, a single `CTRL-L` press is enough, while when it's opened in a normal window one needs to press `CTRL-L` twice.

**To Reproduce**
1. Run `gvim --clean -c "source tmp_1.vim"` with
```vim
" tmp_1.vim
call term_start(&shell, #{hidden: 1, term_finish: 'close'})
    \ ->popup_create(#{minwidth: 70, minheight: 10, border: []})
set wincolor=Pmenu
```
2. Press `CTRL-L`. Now the color of the terminal is set correctly to `wincolor`.
3. The same can be observed in terminal Vim when `termguicolors` is enabled. Run `vim --clean -c "source tmp_2.vim"`:
```vim
" tmp_2.vim
set termguicolors
call term_start(&shell, #{hidden: 1, term_finish: 'close'})
    \ ->popup_create(#{minwidth: 70, minheight: 10, border: []})
set wincolor=Pmenu
```
With `notermguicolors` the color of the terminal window is set correctly.

**Expected behavior**
Vim should set `wincolor` immediately without having to press `CTRL-L`, just like with `notermguicolors` in terminal Vim.

**Screenshots**
Gvim: before (left) and after (right) pressing `CTRL-L`:

![out](https://user-images.githubusercontent.com/6266600/77183107-1f59af80-6ace-11ea-9ad4-522eb2cdd103.png)

**Environment**
- Vim version `8.2.416`
- Terminal: xfce4-terminal, xterm
- Gentoo Linux

**Comment by lacygoill** [comment]
I can't reproduce on 8.2.418, neither in gVim, nor in xterm; the background of the popup window is highlighted according to `Pmenu` almost immediately (although it's briefly highlighted in white before that).

However, when I do the same experiment in st instead of xterm, then `Pmenu` is not used, even after a manual redraw.
If I start Vim by setting `TERM` to `xterm`, then I get back the behavior observed in xterm; i.e. `Pmenu` is used immediately:

    TERM=xterm vim --clean -c "source tmp_2.vim"
    ^^^^^^^^^^

So the issue is probably influenced by the current value of `TERM`, and maybe something else...

**Comment by lacygoill** [comment]
> and maybe something else...

Maybe some terminal setting as displayed by `:pu=execute('set termcap')`.

**Comment by lacygoill** [comment]
> However, when I do the same experiment in st instead of xterm, then Pmenu is not used, even after a manual redraw.

That's because `t_8f` and `t_8b` were not set then:

    let &t_8f = "\e[38;2;%lu;%lu;%lum"
    let &t_8b = "\e[48;2;%lu;%lu;%lum"

**Comment by brammool** [comment]
Strange.  I can make it work with a timer:

``` vim
func Redraw(id)
  call feedkeys("\<C-L>", 't')
endfunc

call timer_start(500, function('Redraw'))
```

But using ":redraw!" instead of feedkeys() doesn't work.

**Comment by bfrg** [comment]
I was pressing `CTRL-L` while in terminal insert-mode. So it's not even Vim's `:redraw`. Running `clear` in the shell works, too.

> But using ":redraw!" instead of feedkeys() doesn't work.

I have also noticed that popup windows are not consistently updated when the cursor stays in command-line mode. For example, a `popup-filter` can be called from command-mode but the popup window is never redrawn. I can open a separate issue about that.

I'm not sure if this is relevant but this is the output of `set termcap?` in Gvim:
```
--- Terminal codes ---
  t_AB=               t_fs=               t_RT=               t_u7=
  t_AF=               t_GP=               t_RV=               t_ue=^[|8H
  t_AL=^[|%p1%dI      t_IE=               t_Sb=               t_us=^[|8h
  t_al=^[|i           t_IS=               t_SC=               t_ut=y
  t_bc=               t_ke=               t_se=^[|16H         t_vb=^[|f
  t_BE=               t_ks=               t_Sf=               t_ve=
  t_BD=               t_le=^H             t_SH=               t_vi=
  t_cd=               t_mb=               t_SI=               t_VS=
  t_ce=^[|$           t_md=^[|2h          t_Si=               t_vs=
  t_cl=^[|C           t_me=^[|31H         t_so=^[|16h         t_WP=
  t_Ce=^[|8C          t_mr=^[|1h          t_SR=               t_WS=
  t_Co=               t_ms=y              t_sr=               t_xn=y
  t_CS=               t_nd=^L             t_ST=               t_xs=
  t_Cs=^[|8c          t_op=               t_Te=^[|4C          t_ZH=^[|4h
  t_da=               t_RF=               t_te=               t_ZR=^[|4H
  t_db=               t_RB=               t_TE=               t_8f=
  t_DL=^[|%p1%dD      t_RC=               t_ti=               t_8b=
  t_dl=^[|d           t_RI=               t_TI=
  t_EC=               t_Ri=               t_Ts=^[|4c
  t_EI=               t_RS=               t_ts=
  t_cm=^[|%p1%d;%p2%dM
  t_cs=^[|%p1%d;%p2%dR
  t_CV=^[|%p1%d;%p2%dV
```

**Comment by lacygoill** [comment]
Not sure it's relevant, but how do you configure the binary?

With this configuration, I can't reproduce on Ubuntu 16.04:

    ./configure  \
      --enable-cscope                \
      --enable-fail-if-missing       \
      --enable-gui=gtk2              \
      --enable-luainterp=dynamic     \
      --enable-multibyte             \
      --enable-python3interp=dynamic \
      --enable-terminal              \
      --prefix=/usr/local            \
      --with-compiledby=user         \
      --with-features=huge           \
      --with-luajit

**Comment by bfrg** [comment]
I'm running `./configure` without any options. 

This is the output of `version`:
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Mar 20 2020 20:17:41)
Included patches: 1-420
Compiled by bfrg@laptop
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_old_static
+arabic            +file_in_path      +mouse_urxvt       -tag_any_white
+autocmd           +find_in_path      +mouse_xterm       -tcl
+autochdir         +float             +multi_byte        +termguicolors
-autoservername    +folding           +multi_lang        +terminal
+balloon_eval      -footer            -mzscheme          +terminfo
+balloon_eval_term +fork()            +netbeans_intg     +termresponse
+browse            +gettext           +num64             +textobjects
++builtin_terms    -hangul_input      +packages          +textprop
+byte_offset       +iconv             +path_extra        +timers
+channel           +insert_expand     -perl              +title
+cindent           +job               +persistent_undo   +toolbar
+clientserver      +jumplist          +popupwin          +user_commands
+clipboard         +keymap            +postscript        +vartabs
+cmdline_compl     +lambda            +printer           +vertsplit
+cmdline_hist      +langmap           +profile           +virtualedit
+cmdline_info      +libcall           -python            +visual
+comments          +linebreak         -python3           +visualextra
+conceal           +lispindent        +quickfix          +viminfo
+cryptv            +listcmds          +reltime           +vreplace
+cscope            +localmap          +rightleft         +wildignore
+cursorbind        -lua               -ruby              +wildmenu
+cursorshape       +menu              +scrollbind        +windows
+dialog_con_gui    +mksession         +signs             +writebackup
+diff              +modify_fname      +smartindent       +X11
+digraphs          +mouse             +sound             -xfontset
+dnd               +mouseshape        +spell             +xim
-ebcdic            +mouse_dec         +startuptime       +xpm
+emacs_tags        +mouse_gpm         +statusline        +xsmp_interact
+eval              -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary        
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/tmp/vim-install/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib64/gtk-2.0/include -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/lib64/libffi/include -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2   -g -O2 -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11   -lm -ltinfo -lelf -lnsl  -lcanberra -lacl -lattr -lgpm -ldl
```

**Comment by lacygoill** [comment]
I guess some system library is influencing the issue.  I can definitely see the white background, but only for a fraction of a second:
![gif](https://user-images.githubusercontent.com/8505073/77210381-2948d600-6b01-11ea-962e-3b88180dd7ea.gif)

**Comment by brammool** [comment]
It appears the trick is to make the shell trigger the redraw.  From the Vim side, and specifically libvterm, the colors need to be set before the program in the terminal draws anything.  Once the terminal is open, changing the colors won't have an immediate effect, since Vim and libvterm can't tell where the default color was used.

To avoid the problem, set the Terminal highlighting before creating the terminal:
hi link Terminal Pmenu
call term_start(&shell, #{hidden: 1, term_finish: 'close'})
    \ ->popup_create(#{minwidth: 70, minheight: 10, border: []})

I don't see a way to redraw the terminal without destroying what the program running in the terminal has already displayed.

**Comment by bfrg** [comment]
Would it be possible to add a new `highlight` entry to `term_start()` that specifies the `Terminal` color for that specific terminal buffer? So something like:
```vim
call term_start(&shell, #{hidden: 1, highlight: 'Pmenu'})
```
If not specified it falls back to `Terminal` which itself falls back to `Normal` when not set by a colorscheme.

`term_start()` already has the entry `ansi_colors` for specifying 16 color names. Adding a new `highlight` entry that sets the foreground/background color makes it complete.
